content/6_Advanced/Treaps.mdx

```diff
@@ -1,27 +1,128 @@
+
+![Treap](./treap.png)
+
+Treaps have two main operations: splitting and merging. Other operations like insert, erase, and searching can be implemented in terms of these operations. 
+
 ## Splitting
 
+The `split` method takes in a pointer to the root of a treap $\texttt{root}$ and a value $x$, and returns two treaps denoted as $\texttt{left}$ and $\texttt{right}$. Like the name suggests, it splits the tree such that all nodes in $\texttt{left}$ have keys less than or equal to $x$ and all nodes in $\texttt{right}$ have keys greater than $x$. 
+
+We can now implement it recursively. Let the left child of a node $n$ be denoted as $n.\texttt{left}$ and the right child as $n.\texttt{right}$.
+- If $\texttt{root} \leq x$, then both the root and the left subtree belong to $\texttt{left}$. We now consider a call to `split` on $\texttt{root.right}$ and note its results as $\texttt{left'}$ and $\texttt{right'}$. Finally, $\texttt{left}$ contains $\texttt{left'}$ and $\texttt{right} = \texttt{right'}$
+- If $\texttt{root} > x$, then both the root and the right subtree belong to $\texttt{right}$. We now consider a call to `split` on $\texttt{root.left}$ and note its results as $\texttt{left'}$ and $\texttt{right'}$. Finally, $\texttt{right}$ contains $\texttt{right'}$ and $\texttt{left} = \texttt{left'}$
+
 ## Merging
 
-## Merging (Arbitrary Keys)
+The `merge` method inverts the `split` method by taking in two treaps $\texttt{left}$ and $\texttt{right}$ and returns a single treap that has the nodes of both treaps. It works under the assuption that all keys $x \in \texttt{left}$ are strictly smaller than all keys $y \in \texttt{right}$. Furthermore, we need to merge these two treaps such that the resultant treap still satisfies the max heap property. 
+
+We root the resultant treap at the root node that has the higher priority, and recursively call `merge` on the other tree and the corresponding subtree of the chosen tree. 
+## Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+struct Node {
+	int val, pri;
+	Node *left, *right;
 
-<Problems problems="treeRot" />
+	Node(int _val) : val(_val), pri(rand()), left(NULL), right(NULL) {};
+} *root;
+
+void split(Node *root, int x, Node &*left, Node &*right) {
+	if (!root) {
+		left = right = NULL;
+		return;
+	}
+
+	if (root->val <= x) {
+		split(root->right, x, root->right, right);
+		left = root;
+	}
+	else {
+		split(root->left, x, left, root->left);
+		right = root;
+	}
+}
+
+void merge(Node &*root, Node *left, Node *right) {
+	if (!l || !r) {
```

Comment:
uuuuuuhhhh
where are `l` and `r` defined
i don't see them in the global scope
and they aren't in any of the arguments

Reply:
also comments explaining what's actually happening would be nice
or i'm just dumb

Reply:
i may or may not have copied code with different variable names that I forgot to change